app/controllers/registrations_controller.rb

```diff
@@ -3,8 +3,11 @@ class RegistrationsController < Devise::RegistrationsController
 
   def new
     if user_signed_in?
-      redirect_to dashboard_path
+      redirect_to root_path(signin: "true")
     else
+      if URI(request.referer || "").host == URI(request.base_url).host
+        store_location_for(:user, request.referer)
+      end
```

Comment:
Makes use of Devise's `store_location_for` to redirect the users back using the referer (if valid)